### ISSUE TITLE
problems with numpy

### ISSUE THREAD

**Issue by CLG68** [HBioquant/DiffBind#4]
Hi,

I reinstalled DiffBindDock on a new PC using env.yaml.  Nothing in this section was installed:
  - pip:
    - --extra-index-url https://download.pytorch.org/whl/cu117
    - --find-links https://pytorch-geometric.com/whl/torch-1.13.1+cu117.html
    - scikit-learn==1.4.1
    - scipy==1.12.0
    - torch-cluster==1.6.0+pt113cu117
    - torch-scatter==2.1.0+pt113cu117
    - torch-sparse==0.6.16+pt113cu117
    - torch-spline-conv==1.2.1+pt113cu117
    - torchmetrics==0.11.0
    - torch-geometric==2.2.0
    
    So I manually installed them:
    
    pip install torch-cluster==1.6.0+pt113cu117 torch-scatter==2.1.0+pt113cu117 torch-sparse==0.6.16+pt113cu117 torch-spline-conv==1.2.1+pt113cu117 -f https://pytorch-geometric.com/whl/torch-1.13.1+cu117.html
    
pip install torch-geometric==2.2.0

installing torch-geometric also installed: scikit-learn==1.4.1 and changed joblib to 1.3.2.  I could not revert joblib to 1.1.0 as it is not compatible with this version of torch-geometric.  One thing I did not do is to downgrade torchmetrics to 0.11.0 before installing torch-geometric, I left it to version 1.3.2.

Torch looks of with cuda:

(diffbindfr) christian@Linux00:/media/christian/VS1/VS/VS_DiffBindFR$ python
Python 3.9.18 | packaged by conda-forge | (main, Dec 23 2023, 16:33:10) 
[GCC 12.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.__version__
'1.13.1'
>>> torch.cuda.is_available()
True
>>> torch.version.cuda
'11.7'
>>> exit ()

This works also fine after I reinstalled torch-geometric:
python DiffBindFR/app/predict.py -h

However, I get these errors trying to run one of the command in the read.me file:
I do not know if all those errors are coming from the problem with numpy version.  However, I cannot downgrade numpy below 1.24.0 as the min version for different packages is between 1.20 and 1.24.

(diffbindfr) christian@Linux00:/media/christian/VS1/VS/VS_DiffBindFR/DiffBindFR/app$ python predict.py -l ../../examples/forward/mols -p ../../examples/forward/3dbs_protein.pdb -o ./test -np 40 -gpu 0 -cpu 6 -bs 6 -n forward
/home/christian/anaconda3/envs/diffbindfr/lib/python3.9/site-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.
  warnings.warn(
 ____  _  __  __ ____  _           _ _____ ____  
|  _ \(_)/ _|/ _| __ )(_)_ __   __| |  ___|  _ \ 
| | | | | |_| |_|  _ \| | '_ \ / _` | |_  | |_) |
| |_| | |  _|  _| |_) | | | | | (_| |  _| |  _ < 
|____/|_|_| |_| |____/|_|_| |_|\__,_|_|   |_| \_\ 
2024-03-21 00:15:05,051 - DiffBindFR - INFO - Total loaded jobs: 15.
2024-03-21 00:15:05,051 - DiffBindFR - INFO - Job Slice Info: (0, 15).
2024-03-21 00:15:05,051 - DiffBindFR - INFO - Running jobs: 15.
2024-03-21 00:15:05,054 - DiffBindFR - INFO - Start to prepare job (experiment name: forward).
INFO: Pandarallel will run on 6 workers.
INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.
2024/03/21 00:15:05 - InferenceDataset - **ERROR - module 'numpy' has no attribute 'int'.**
`np.int` was a deprecated alias for the builtin `int`. To avoid this error in existing code, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
The aliases was originally deprecated in NumPy 1.20; for more details and guidance see the original release note at:
    https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations from /media/christian/VS1/VS/VS_DiffBindFR/examples/forward/3dbs_protein.pdb from crystal ligand /media/christian/VS1/VS/VS_DiffBindFR/DiffBindFR/app/../../examples/forward/3dbs_protein_crystal.sdf
Use Background Generator supported dataloader.
2024-03-21 00:15:05,606 - DiffBindFR - INFO - dock Status: Prep task is Done!
Initializing diffusion model...
2024-03-21 00:15:07,631 - DiffBindFR - INFO - load checkpoint from local path: /media/christian/VS1/VS/VS_DiffBindFR/DiffBindFR/weights/diffbindfr_paper.pth
2024-03-21 00:15:08,466 - DiffBindFR - INFO - Running model inference...
[                                                  ] 0/560, elapsed: 0s, ETATraceback (most recent call last):
  File "/media/christian/VS1/VS/VS_DiffBindFR/DiffBindFR/app/predict.py", line 265, in <module>
    sys.exit(main())
  File "/media/christian/VS1/VS/VS_DiffBindFR/DiffBindFR/app/predict.py", line 260, in main
    runner(df, args)
  File "/media/christian/VS1/VS/VS_DiffBindFR/DiffBindFR/app/predict.py", line 131, in runner
    pairs_results = common.inferencer(
  File "/media/christian/VS1/VS/VS_DiffBindFR/DiffBindFR/common/engines.py", line 204, in inferencer
    model_out = model_run(dl, model, show_traj)
  File "/media/christian/VS1/VS/VS_DiffBindFR/DiffBindFR/common/engines.py", line 174, in model_run
    for idx, data in enumerate(dl):
  File "/home/christian/anaconda3/envs/diffbindfr/lib/python3.9/site-packages/prefetch_generator/__init__.py", line 116, in next
    raise next_item
  File "/home/christian/anaconda3/envs/diffbindfr/lib/python3.9/site-packages/prefetch_generator/__init__.py", line 98, in run
    for item in self.generator: self.queue.put((True , item))
  File "/home/christian/anaconda3/envs/diffbindfr/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 628, in __next__
    data = self._next_data()
  File "/home/christian/anaconda3/envs/diffbindfr/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 1333, in _next_data
    return self._process_data(data)
  File "/home/christian/anaconda3/envs/diffbindfr/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 1359, in _process_data
    data.reraise()
  File "/home/christian/anaconda3/envs/diffbindfr/lib/python3.9/site-packages/torch/_utils.py", line 543, in reraise
    raise exception
ValueError: Caught ValueError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/home/christian/anaconda3/envs/diffbindfr/lib/python3.9/site-packages/torch/utils/data/_utils/worker.py", line 302, in _worker_loop
    data = fetcher.fetch(index)
  File "/home/christian/anaconda3/envs/diffbindfr/lib/python3.9/site-packages/torch/utils/data/_utils/fetch.py", line 58, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/home/christian/anaconda3/envs/diffbindfr/lib/python3.9/site-packages/torch/utils/data/_utils/fetch.py", line 58, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/media/christian/VS1/VS/VS_DiffBindFR/druglib/datasets/custom_dataset.py", line 263, in __getitem__
    data = self.get(idx)
  File "/media/christian/VS1/VS/VS_DiffBindFR/druglib/datasets/custom_dataset.py", line 326, in get
    return self._prepare_test_sample(self.indices[idx])
  File "/media/christian/VS1/VS/VS_DiffBindFR/DiffBindFR/common/inference_dataset.py", line 583, in _prepare_test_sample
    mdl_inp = getattr(self.PairData, obj + 's')[row[name]].model_input
  File "/media/christian/VS1/VS/VS_DiffBindFR/druglib/datasets/lmdbdataset.py", line 67, in __getitem__
    raise ValueError(f'query index {idx.decode()} not in lmdb.')
ValueError: query index 3dbs_protein not in lmdb.

**Comment by HBioquant** [comment]
Hi, Christian,

I apologize for the inconvenience caused by the package version management. The error `ValueError: query index 3dbs_protein not in lmdb.` is caused by the failed protein featurization due to the numpy error `ERROR - module 'numpy' has no attribute 'int'.**np.intwas a deprecated alias for the builtinint.`. Actually, you can pass the argument `--debug` to detect it. 

It is smart to install manually when you found that pytorch-geometric's operator dependencies were missing.

I've uploaded a reference txt file that includes the specific versions of the packages that are known to work well together.  You can find [this file in the repository](https://github.com/HBioquant/DiffBindFR/blob/main/requirements/requirements_reference.txt).  Please try to install the packages using the versions specified in this file.

Though packages like torch-geometric may have stricter version control, downgrading these packages (joblib, numpy, etc) should not cause any issues in my personal tests.

Please let me know if this resolves your issue or if you have any further questions.

Best regards,
Jintao Zhu

**Comment by CLG68** [comment]
Dear Dr Zhu,
Thank you very much for the list.  I see in that list that numpy is > 1.20 so I'm not too sure why I get this message about *ERROR - module 'numpy' has no attribute 'int'.**np.int was a deprecated alias for the built in int "
Would it be possible to create the list using: $conda list -n diffbindfr --export > something.txt , please,
as it would also contain the build and I should be able to use it to create the environment.

Best,
Christian

**Comment by HBioquant** [comment]
Dear Christian,

Thank you for your message. I have read your email and understand your concern. The error you're encountering, "*ERROR - module 'numpy' has no attribute 'int'.**np.int was a deprecated alias for the built in int", is indeed related to the version of numpy. This error typically occurs when using numpy version >=1.24.0. Please see details in the [numpy 1.24.0 document](https://github.com/numpy/numpy/pull/22607).

In the reference version list I provided, the numpy version is <1.24.0, specifically 1.23.5, which should not cause this error. I would recommend trying to install numpy with this version by `pip uninstall numpy` and then `pip install numpy==1.23.5` to see if it resolves your issue.

```
>>> import numpy as np
>>> np.__version__
'1.23.5'
>>> np.int
<stdin>:1: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
<class 'int'>
```

```
>>> import numpy as np
>>> np.__version__
'1.24.0'
>>> np.int
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/lhlai_pkuhpc/lustre2/zhujt/tools/anaconda3/envs/diffbindfr/lib/python3.9/site-packages/numpy/__init__.py", line 284, in __getattr__
    raise AttributeError("module {!r} has no attribute "
AttributeError: module 'numpy' has no attribute 'int'
```

Regarding your request to generate the list using `$conda list -n diffbindfr --export > something.txt`, I will do that and share my conda environment with you in a follow-up email soon. This file will contain the build information as well, which should help you in creating the environment.

Please let me know if you have any further questions or if there's anything else I can assist you with.

Best regards,
Jintao Zhu

**Comment by CLG68** [comment]
Dear Dr Zhu,

Thank you very much for the info on Numpy and also the file.  Downgrading Numpy to 1.23.5 did the trick.  It went much further and generated all the poses.  Eventually, it hits an other set of errors, I guess at the step of scoring the poses.  I will install, on a different PC, a new DiffBindFR using the generated list from your environment...maybe it will solve all the problems.  Even w/o the scores, I reach the step I wanted.  I could always rescore with some other tools.  

Here are the new errors:

(diffbindfr) christian@Linux00:/media/christian/VS1/VS/VS_DiffBindFR/DiffBindFR/app$ python predict.py -l ../../examples/forward/mols -p ../../examples/forward/3dbs_protein.pdb -o ./test -np 40 -gpu 0 -cpu 6 -bs 6 -n forward
/home/christian/anaconda3/envs/diffbindfr/lib/python3.9/site-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.
  warnings.warn(
 ____  _  __  __ ____  _           _ _____ ____  
|  _ \(_)/ _|/ _| __ )(_)_ __   __| |  ___|  _ \ 
| | | | | |_| |_|  _ \| | '_ \ / _` | |_  | |_) |
| |_| | |  _|  _| |_) | | | | | (_| |  _| |  _ < 
|____/|_|_| |_| |____/|_|_| |_|\__,_|_|   |_| \_\ 
2024-03-21 23:49:41,749 - DiffBindFR - INFO - Total loaded jobs: 15.
2024-03-21 23:49:41,749 - DiffBindFR - INFO - Job Slice Info: (0, 15).
2024-03-21 23:49:41,749 - DiffBindFR - INFO - Running jobs: 15.
2024-03-21 23:49:41,752 - DiffBindFR - INFO - Start to prepare job (experiment name: forward).
INFO: Pandarallel will run on 6 workers.
INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.
Use Background Generator supported dataloader.
2024-03-21 23:49:47,890 - DiffBindFR - INFO - dock Status: Prep task is Done!
Initializing diffusion model...
2024-03-21 23:49:49,884 - DiffBindFR - INFO - load checkpoint from local path: /media/christian/VS1/VS/VS_DiffBindFR/DiffBindFR/weights/diffbindfr_paper.pth
2024-03-21 23:49:50,637 - DiffBindFR - INFO - Running model inference...
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 201/201 [00:55<00:00,  3.61it/s]
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 201/201 [01:04<00:00,  3.10it/s]
[>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>] 600/600, 0.3 task/s, elapsed: 1729s, ETA:     0s2024-03-22 00:18:39,637 - DiffBindFR - INFO - Model inference is done which tasks 1728.9997351169586s
2024-03-22 00:18:39,637 - DiffBindFR - INFO - Export model results to path: /media/christian/VS1/VS/VS_DiffBindFR/DiffBindFR/app/test/forward/results/model_output.pt
2024-03-22 00:18:39,871 - DiffBindFR - INFO - Export binding structures....
2024-03-22 00:18:56,508 - DiffBindFR - INFO - Binding structure export is completed.
INFO: Pandarallel will run on 6 workers.
INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.
2024-03-22 00:18:56,514 - DiffBindFR - INFO - Start to correct error...
Traceback (most recent call last):
  File "/media/christian/VS1/VS/VS_DiffBindFR/DiffBindFR/app/predict.py", line 265, in <module>
    sys.exit(main())
  File "/media/christian/VS1/VS/VS_DiffBindFR/DiffBindFR/app/predict.py", line 260, in main
    runner(df, args)
  File "/media/christian/VS1/VS/VS_DiffBindFR/DiffBindFR/app/predict.py", line 162, in runner
    pd_df['smina_score'] = pd_df['docked_lig'].parallel_apply(
  File "/home/christian/anaconda3/envs/diffbindfr/lib/python3.9/site-packages/pandarallel/core.py", line 324, in closure
    return wrapped_reduce_function(
  File "/home/christian/anaconda3/envs/diffbindfr/lib/python3.9/site-packages/pandarallel/core.py", line 199, in closure
    return reduce_function(dfs, extra)
  File "/home/christian/anaconda3/envs/diffbindfr/lib/python3.9/site-packages/pandarallel/data_types/series.py", line 34, in reduce
    return pd.concat(datas, copy=False)
  File "/home/christian/anaconda3/envs/diffbindfr/lib/python3.9/site-packages/pandas/core/reshape/concat.py", line 382, in concat
    op = _Concatenator(
  File "/home/christian/anaconda3/envs/diffbindfr/lib/python3.9/site-packages/pandas/core/reshape/concat.py", line 445, in __init__
    objs, keys = self._clean_keys_and_objs(objs, keys)
  File "/home/christian/anaconda3/envs/diffbindfr/lib/python3.9/site-packages/pandas/core/reshape/concat.py", line 504, in _clean_keys_and_objs
    objs_list = list(objs)
  File "/home/christian/anaconda3/envs/diffbindfr/lib/python3.9/site-packages/pandarallel/core.py", line 195, in <genexpr>
    get_dataframe_and_delete_file(output_file_path)
  File "/home/christian/anaconda3/envs/diffbindfr/lib/python3.9/site-packages/pandarallel/core.py", line 189, in get_dataframe_and_delete_file
    data = pickle.load(file_descriptor)
EOFError: Ran out of input

**Comment by HBioquant** [comment]
Please check if your `druglib/ops/smina/smina.static` is functioning properly. Typically, this error originates from your computer being unable to run `Smina` correctly.

**Comment by CLG68** [comment]
None of the prg in the opt folder, including Smina.static had the necessary permission to run.  I gave the permission to execute to the full DiffBindFR folder with sudo chmod -R +x
Now when I run python predict.py -l ../../examples/forward/mols -p ../../examples/forward/3dbs_protein.pdb -o ./test -np 20 -gpu 0 -cpu 6 -bs 6 -n forward
I do not have the smina-related errors and I get the scores.  However, the script hangs forever there waiting for something as I do not see cpu resources being used.  In the results folder, I have these files: model_output.pt, results.csv, results_ec.csv, results_ec_smina_top1.csv .  Isn't it supposed to use 2 different scoring functions?  I see only smina's results.

This is where it blocks:
2024-03-22 09:31:59,599 - DiffBindFR - INFO - Start to correct error...
2024-03-22 09:32:16,856 - DiffBindFR - INFO - Error correction is completed.
2024-03-22 09:32:16,861 - DiffBindFR - INFO - Start to MDN based pose scoring...
Processing 180 jobs, chunk 50000 per iter, total 1 iters...:   0%| | 0/1 [00:00<INFO: Pandarallel will run on 6 workers.
INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.

Also, even if I set -cpu 6 , it never use more than 1 thread at a time... I'm not sure it works in parallel, maybe it comes from the dependencies.  It is difficult for me to evaluate the difference between using more of less cpu resources as it does not go through the full process.  As I would like to evaluate it is screening, the faster the better.

Best

**Comment by HBioquant** [comment]
Hi. Your current progress is being scored using mdn. Once it is done it will get a results_ec_mdn_top1.csv. 

As for smina bin, you can find it [here](https://github.com/HBioquant/DiffBindFR/blob/main/druglib/ops/smina/). But, i see it seems work for you now. 

Here, I show you the complete output running on my machine.

![image](https://github.com/HBioquant/DiffBindFR/assets/68720788/4427cc73-4b5f-48cc-af14-b7ccf680cb4b)

It's strange that under the setting of 20 poses, only a total of 180 poses (maybe there are 120 poses missing?) were generated (as you showed `Processing 180 jobs`), theoretically, there should be 300 poses generated. I suggest you change the experiment name (set `-n forward_bug`) and redo the experiment, first set `-np 1` for testing, you can also further set -e to control the number of pairs that need to be docked (such as `-e 2`). I will further look forward to a normal output from you.

The complete debug code:
`
python predict.py -l ../../examples/forward/mols -p ../../examples/forward/3dbs_protein.pdb -o ./test -np 1 -e 2 -gpu 0 -cpu 1 -bs 6 -n forward_debug --debug
`

**Comment by CLG68** [comment]
Hi,

For the 20 poses vs 180, I might have recopied the code from one I tested different number of poses.  I think it is ok.

The debug code works, even if I increase the number of poses but if I change -cpu 1 to -cpu 3 than it hangs there at the scoring step.  It could have been related to Pandarallel as it does not matter how many threads I assigned, DiffBindFR does not use more than one at a time.  However, I have the same version 1.6.3 and also same version of Pandas.  I forced Pandarallel to 1.6.5 and it did not solved the problem.
I'll compare all the version of everything in my environment to your list... maybe I'll find something and I will get back to you.

I really appreciate your help with this.
Best

**Comment by HBioquant** [comment]
Now it seems that the problem lies with Pandarallel, which may not be applicable in your system. You can refer to [this issue](https://github.com/nalepae/pandarallel/issues/122) to try to solve it. If it still doesn't solve your problem, you can actually replace [Pandarallel](https://github.com/HBioquant/DiffBindFR/blob/main/DiffBindFR/scoring/dataset/inference.py#L156) to handle multi-process operation, such as using joblib or the more basic multiprocessing, which are very easy to replace in my code.

Thank you very much for testing the code. Although I have tested it several times locally without any problems, it made me realize that there are many inconveniences when my code is open for others to use. I will improve them in the future!
Best

**Comment by CLG68** [comment]
Dear Dr Zhu,

I completely reinstalled DiffBindFR with an environment.yaml file I created based on the list you gave me. I installed the torch-related files manually + downgraded  numpy to 1.23.5...  Now everything seems to work fine.  Thank you very much for your precious help and also for creating this tool.  I'll attach 2 files, my environment.yaml file and the list of modifications I did manually, just in case somebody else end up with the same problem.

Best,
Christian
[environment.zip](https://github.com/HBioquant/DiffBindFR/files/14730278/environment.zip)

[environment-complement.txt](https://github.com/HBioquant/DiffBindFR/files/14730268/environment-complement.txt)

**Comment by Daisyli95** [comment]
Thanks a lot for you posting your discussions, which really help people like me meeting the same problems

**Comment by CLG68** [comment]
@Daisyli95  I'm using now numpy 1.23.2; it is more stable than with 1.23.5

I created a script to install the environment. This environment works well with smina but not mdn so I use the argument: --no_mdn_scoring .  I did not install the requirements for pl.py as for me, this does not work... for ex: t uses unicon and its reference is hard-coded in the script: exec_bin = '/data02/zhujintao/tools/zbh_tools/unicon_1.4.2/unicon'.

I attached an updated installation script (04/15/2025) that works for me.  However, I'm not using pl.py nor mdn.

[install_DiffBindFR.zip](https://github.com/user-attachments/files/19749186/install_DiffBindFR.zip)
